builder/dispatchers.go

```diff
@@ -226,10 +226,16 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 	// set Cmd manually, this is special case only for Dockerfiles
 	b.Config.Cmd = config.Cmd
 	runconfig.Merge(b.Config, config)
+	// set build-time environment for 'run'. We let dockerfile
+	// environment override build-time environment.
+	env := b.Config.Env
```

Comment:
if I'm following this correctly, you're adding the env vars that the user set on the cmd line when they issued the "docker build".  But, you're only adding it to the BuildContext for the RUN command.  Why not add it to the BuildContext's env so it applies to all commands?  Granted RUN is probably the only one that will directly use it, but by making it global to the build context you can then use those env vars for variable substitutions in other commands.  For example, if someone does:
ADD $foo /tmp/
we'll expand $foo - but I don't think your code will work for this case.  Do I have that right?